File: Projeto1/src/project1.f90
PROGRAM project1

	CALL get_samples(1,8)
	CALL system("mv fort.1 ../data/Fortran_RC.csv")
	CALL get_samples(2,8)
	CALL system("mv fort.1 ../data/Fortran_CR.csv")

END PROGRAM

! Get the samples
SUBROUTINE get_samples(implementation_version, bytes_per_element)

	INTEGER, INTENT(IN) :: implementation_version
	INTEGER, INTENT(IN) :: bytes_per_element

	! Declaring some variables;
	CHARACTER(30) :: line_content

	REAL(8), ALLOCATABLE :: A(:,:)
	REAL(8), ALLOCATABLE :: x(:)
	REAL(8), ALLOCATABLE :: b(:)
	
	INTEGER A_allocationStat, x_allocationStat, b_allocationStat

	INTEGER :: nmax, n, size_nElements;
	REAL(8) :: t1, t2, t_elapsed;

	! Getting the maximum value of n
	CALL find_nmax(bytes_per_element, nmax)
	
	! Generating the samples
	n = 1

	DO WHILE(n.LE.nmax)

		size_nElements = bytes_per_element*n

		! Trying to allocate both matrix and the two vectors
		allocate(A(size_nElements,size_nElements), stat=A_allocationStat)
		allocate(x(size_nElements),stat=x_allocationStat)
		allocate(b(size_nElements),stat=b_allocationStat)

		! If all have been done successfully
		IF ((A_allocationStat == 0).AND.(x_allocationStat == 0).AND.(b_allocationStat == 0)) THEN 
			
			! Initialing A, x and b
			CALL squareMatrix_randomInitializer(n, A)
			CALL vector_randomInitializer(n, x)
			CALL vector_zerosInitializer(n, b)

			IF (implementation_version == 1) THEN
				
				! Start time counting
				CALL cpu_time(t1)

				CALL mult_matrixVector_version1(n, A, x, b)

				! Stop clocktime counting
				CALL cpu_time(t2)
		
			ELSE
				
				! Start time counting
				CALL cpu_time(t1)

				CALL mult_matrixVector_version2(n, A, x, b)

				! Stop time counting
				CALL cpu_time(t2)
			END IF

			t_elapsed = (t2-t1) * (10**6)

			! Writing sample to file
			WRITE(line_content,"(I10.1,A1,I15)") n,",",int(t_elapsed)
			WRITE(1,*) line_content

			! Free allocated memory
			IF (A_allocationStat == 0) THEN
				DEALLOCATE(A)
			END IF
			
			IF (x_allocationStat == 0) THEN
				DEALLOCATE(x)
			END IF

			IF (b_allocationStat == 0) THEN
				DEALLOCATE(b)
			END IF

		! Otherwise
		ELSE
			
			! Free allocated memory
			IF (A_allocationStat == 0) THEN
				DEALLOCATE(A)
			END IF
			
			IF (x_allocationStat == 0) THEN
				DEALLOCATE(x)
			END IF

			IF (b_allocationStat == 0) THEN
				DEALLOCATE(b)
			END IF

			! Stop loop
			STOP

		END IF

		n = 2*n

	END DO 

END SUBROUTINE

! Square Matrix random initialization
SUBROUTINE squareMatrix_randomInitializer(n, matrix)
	
	INTEGER, INTENT(IN) :: n
	REAL(8), DIMENSION(n,n), INTENT(OUT) :: matrix
	DO i = 1,n
		DO j = 1,n
			matrix(i,j) = RAND()*100
		END DO
	END DO

END SUBROUTINE

! Vector random initialization
SUBROUTINE vector_randomInitializer(n, vector)

	INTEGER, INTENT(IN) :: n
	REAL(8), DIMENSION(n), INTENT(OUT) :: vector
	DO i = 1,n
		vector(i) = RAND()*100
	END DO

END SUBROUTINE

! Initialize vector with zeros
SUBROUTINE vector_zerosInitializer(n, vector)

	INTEGER, INTENT(IN) :: n
	REAL(8), DIMENSION(n), INTENT(OUT) :: vector
	
	vector(:) = 0

END SUBROUTINE

! Matrix-vector product, version 1 (outer loop along rows and inner loop along columns)
SUBROUTINE mult_matrixVector_version1(n, matrix, vector, result)

	INTEGER, INTENT(IN) :: n
	REAL(8), DIMENSION(n,n), INTENT(IN) :: matrix
	REAL(8), DIMENSION(n), INTENT(IN) :: vector
	REAL(8), DIMENSION(n), INTENT(OUT) :: result

	! Multiplying matrix by vector and storing it in result
	DO i = 1, n
		DO j = 1, n
			result(i) = result(i) + matrix(i,j) * vector(j);
		END DO
	END DO

END SUBROUTINE

! Matrix-vector product, version 2 (outer loop along columns and inner loop along rows)
SUBROUTINE mult_matrixVector_version2(n, matrix, vector, result)

	INTEGER, INTENT(IN) :: n
	REAL(8), DIMENSION(n,n), INTENT(IN) :: matrix
	REAL(8), DIMENSION(n), INTENT(IN) :: vector
	REAL(8), DIMENSION(n), INTENT(OUT) :: result

	! Multiplying matrix by vector and storing it in result
	DO j = 1, n
		DO i = 1, n
			result(i) = result(i) + matrix(i,j) * vector(j);
		END DO
	END DO

END SUBROUTINE

! Find maximum value of parameter n
SUBROUTINE find_nmax(bytes_per_element, nmax)

	INTEGER, INTENT(IN) :: bytes_per_element
	INTEGER, INTENT(OUT) :: nmax
	
	! Defining some variables
	INTEGER freeMemoryInBytes, size_nElements
	INTEGER A_allocationStat, x_allocationStat, b_allocationStat

	INTEGER n_ideal
	INTEGER n_current
	INTEGER n_last

	REAL(8), ALLOCATABLE :: A(:,:)
	REAL(8), ALLOCATABLE :: x(:)
	REAL(8), ALLOCATABLE :: b(:)

	! Getting free memory
	CALL system('python3 freeMemory.py')

	OPEN(unit=1,file='./freeMemory.txt')
	READ(1,*) freeMemoryInBytes
	CLOSE(unit=1)

	! Initializing some variables
	n_ideal = (-1 + INT(sqrt(1.0 + REAL(freeMemoryInBytes/bytes_per_element))))
	n_current = (n_ideal/2)
	n_last = 0;

	! Finding maximum value of n
	DO WHILE (n_last /= n_current)

		size_nElements = bytes_per_element*n_current

		! Uncomment the line below to print the current value of n
		! PRINT *, n_current

		! Trying to allocate both matrix and the two vectors
		allocate(A(size_nElements,size_nElements), stat=A_allocationStat)
		allocate(x(size_nElements),stat=x_allocationStat)
		allocate(b(size_nElements),stat=b_allocationStat)

		! If all have been done successfully
		IF ((A_allocationStat == 0).AND.(x_allocationStat == 0).AND.(b_allocationStat == 0)) THEN 
			n_last = n_current
			n_current = (n_ideal + n_current)/2
		! Otherwise
		ELSE
			n_current = (n_last + n_current)/2;
		END IF
		
		! Free allocated memory for the next iteration
		IF (A_allocationStat == 0) THEN
			DEALLOCATE(A)
		END IF
		
		IF (x_allocationStat == 0) THEN
			DEALLOCATE(x)
		END IF

		IF (b_allocationStat == 0) THEN
			DEALLOCATE(b)
		END IF

	END DO
        
        PRINT *, n_current
	nmax = n_current

	RETURN 

END SUBROUTINE

File: Projeto1/src/project1.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <sys/mman.h>
#include <time.h>

int find_nmax(size_t bytes_per_element) {
	/* This function calculates and retuns the maximum value for n for our problem, i.e, we are 
	considering that we need to allocate a square matrix (n x n) and two vectors 
	n-dimensional.
	
	Parameter
	=========
	size_t bytes_per_element: number of bytes allocated per matrix/vector's element
	*/

	// Getting free memory
	char buffer[20];
	size_t freeMemoryInBytes;

	system("python3 freeMemory.py");

	FILE *f;
	f = fopen("./freeMemory.txt", "r"); 
	fscanf(f,"%s",buffer);
	freeMemoryInBytes = (size_t) atoi(buffer);
	fclose(f);

	// Defining some variable
	int n_ideal, n_current, n_last;
	double *A, *x, *b;

	// Initializing some variables
	n_ideal = (int) (-1 + sqrt(1 + (freeMemoryInBytes/bytes_per_element)));
	n_current = (int) (n_ideal/2);
	n_last = 0;

	// Finding maximum value of n
	while (n_last != n_current) {

		// Uncomment the line below to print the current value of n
		// printf("%d\n",n_current);

		// Trying to allocate both matrix and the two vectors
		A = (double*) calloc(bytes_per_element, (size_t) pow(n_current,2));
		x = (double*) calloc(bytes_per_element, (size_t) n_current);
		b = (double*) calloc(bytes_per_element, (size_t) n_current);

		// If all have been done successfully
		if ((A != NULL) && (x != NULL) && (b != NULL)) {
			n_last = n_current;
			n_current = (int) (n_ideal + n_current)/2;
		} 
		// Otherwise
		else {
			n_current = (int) (n_last + n_current)/2;
		}

		// Free allocated memory for the next iteration
		free((void*) A);
		free((void*) x);
		free((void*) b);
	}
	printf("n_max: %d\n", n_current);
	return n_current;
}

void Vector_Zeros(int n, double *vector) {
	/* 
	It initializes the vector passed as reference as an array of zeros.

	Parameter
	=========

	int n: dimension of the vector
	double *vector: array with n entries
	*/

	for (int i = 0; i < n; i++) {
		vector[i] = (double) 0;
	}
}

void Mult_MatrixVector_Version1(int n, double *matrix, double *vector, double *result) {
	/* 
	It calculates the vector 'result' as a multiplication of the matrix by the vector.
	In comparison with Mult_MatrixVector_Version2 function, the only difference is the order of loops.

	Parameter
	=========

	int n: dimension of square matrix A and the vectors
	double matrix[n][n]: bidimensional array, i.e., a square matrix
	double vector[n]: array with n entries
	double result[n]: array with n entries. This vector is used to store the results and should be a
					  vector of zeros.
	*/ 

	// Multiplying matrix by vector and storing it in result
	for (int i = 0; i < n; i++) {

		for (int j = 0; j < n; j++) {
			result[i] += matrix[i*n + j] * vector[j];
		}

	}
}

void Vector_RandomInitializer(int n, double *vector) {
	/* 
	It initializes a vector n-dimensional with random values between 0 and 100. 
	All the coefficients are written directly in the vector passed as reference.

	Parameter
	=========

	int n: vector dimension
	double vector[n]: array with n entries
	*/

	// Setting random seed
	srand(time(NULL));

	// Initializing all the coefficients
	for (int i = 0; i < n; i++) {

		vector[i] = ((double) rand())/RAND_MAX * 100;
		
	}
}

void SquareMatrix_RandomInitializer(int n, double *matrix) {
	/* 
	It initializes a square matrix (n x n) with random values between 0 and 100. 
	All the coefficients are written directly in the matrix passed as reference.

	Parameter
	=========

	int n: matrix dimension
	double matrix[n][n]: bidimensional array, i.e., a square matrix
	*/

	// Setting random seed
	srand(time(NULL));

	// Initializing all the coefficients
	for (int i = 0; i < n; i++) {

		for (int j = 0; j < n; j++) {
			matrix[i*n + j] = ((double) rand())/RAND_MAX * 100;
		}

	}
}

void Mult_MatrixVector_Version2(int n, double *matrix, double *vector, double *result) {
	/* 
	It calculates the vector 'result' as a multiplication of the matrix by the vector.
	In comparison with Mult_MatrixVector_Version1 function, the only difference is the order of loops.

	Parameter
	=========

	int n: dimension of square matrix A and the vectors
	double matrix[n][n]: bidimensional array, i.e., a square matrix
	double vector[n]: array with n entries
	double result[n]: array with n entries. This vector is used to store the results and should be a
					  vector of zeros.
	*/ 

	// Multiplying matrix by vector and storing it in result
	for (int j = 0; j < n; j++) {
		
		for (int i = 0; i < n; i++) {
			result[i] += matrix[i*n + j] * vector[j];
		}

	}
}

void get_samples(int implementation_version, size_t bytes_per_element, char *destiny) {
	/* 

	Parameter
	=========

	int implementation_version: if 1, it uses the first version of matrix-vector product.
								If 2, it uses the second version of matrix-vector product.

	size_t bytes_per_element: number of bytes allocated per matrix/vector's element

	char *destiny: directory + filename. This parameter will define where the 
				   samples will be stored
	*/

	// Defining some variables;
	double *A, *x, *b;
	int nmax, n;
	clock_t t1,t2;
	long t_elapsed;

	// Openning the file where samples will be written and writing columns names
	FILE *samples;
	samples = fopen(destiny, "w");
	fprintf(samples, "n,time(microseconds)\n");

	// Getting the maximum value of n
	nmax = find_nmax(bytes_per_element);
	
	// Generating the samples
	n = 1;
	while(n <= nmax) {

		// Trying to allocate both matrix and the two vectors
		A = (double*) calloc(bytes_per_element, (size_t) pow(n,2));
		x = (double*) calloc(bytes_per_element, (size_t) n);
		b = (double*) calloc(bytes_per_element, (size_t) n);

		// If all have been done successfully
		if ((A != NULL) && (x != NULL) && (b != NULL)) {
			// Initialing A, x and b
			SquareMatrix_RandomInitializer(n, A);
			Vector_RandomInitializer(n, x);
			Vector_Zeros(n, b);

			if (implementation_version == 1) {
				// Start time counting
				t1 = clock();

				Mult_MatrixVector_Version1(n, A, x, b);

				// Stop clocktime counting
				t2 = clock();
			} 
			else {
				// Start time counting
				t1 = clock();

				Mult_MatrixVector_Version2(n, A, x, b);

				// Stop time counting
				t2 = clock();
			}

			t_elapsed = ((double) t2 - t1) / CLOCKS_PER_SEC * pow(10,6); 


			// Writing sample to file
			fprintf(samples, "%d,%ld\n", n, t_elapsed);

			// Free allocated memory
			free((void*) A);
			free((void*) x);
			free((void*) b);

		} else {
			// Free allocated memory
			free((void*) A);
			free((void*) x);
			free((void*) b);

			// Stop loop
			break;
		}

		n = 2*n;
	}

	// Closing samples file
	fclose(samples);
}


void main() {
	get_samples(1, sizeof(double), "../data/C_RC.csv");
	get_samples(2, sizeof(double), "../data/C_CR.csv");
}
